src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/BufferChunk.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Extensions.Caching.Hybrid.Internal;
+
+// Used to convey buffer status; like ArraySegment<byte>, but Offset is always
+// zero, and we use the most significant bit of the length (usually the sign flag,
+// but we do not need to support negative length) to track whether or not
+// to recycle this value.
+internal readonly struct BufferChunk
+{
+    private const int FlagReturnToPool = (1 << 31);
+
+    private readonly int _lengthAndPoolFlag;
+
+    public byte[]? Array { get; } // null for default
+
+    public int Length => _lengthAndPoolFlag & ~FlagReturnToPool;
+
+    public bool ReturnToPool => (_lengthAndPoolFlag & FlagReturnToPool) != 0;
+
+    public byte[] ToArray()
+    {
+        var length = Length;
+        if (length == 0)
+        {
+            return [];
+        }
+
+        var copy = new byte[length];
+        Buffer.BlockCopy(Array!, 0, copy, 0, length);
+        return copy;
+
+        // Note on nullability of Array; the usage here is that a non-null array
+        // is always provided during construction, so the only null scenario is for default(BufferChunk).
+        // Since the constructor explicitly accesses array.Length, any null array passed to the constructor
+        // will cause an exception, even in release (the Debug.Assert only covers debug) - although in
+        // reality we do not expect this to ever occur (internal type, usage checked, etc). In the case of
+        // default(BufferChunk), we know that Length will be zero, which means we will hit the [] case.
```

Comment:
I figured what triggered and confused me in this class - the .ctor is interleaved with other methods. It's customary (and per the old .NET code guidelines) to place the .ctor after fields and events, and I'm so used to that - and I totally missed the .ctor and, thus, got confused by the `!` operator use.

Reply:
reordering .ctor first